### ISSUE TITLE
per-policy proxy setting in Elastic Agent in Fleet mode

### ISSUE THREAD

**Issue by gbanasiak** [elastic/beats#29542]
**Describe the enhancement:**

Agent connects to 3 elements: artifacts, Fleet Server and Elasticsearch. Today Fleet allows to specify proxy settings in global Fleet settings (Fleet UI) for Elasticsearch output only. To fully support proxy in Fleet mode, Fleet UI should allow to specify proxy settings for _all_ agent connections (not only the ES). Furthermore Fleet should allow to specify proxy settings per agent policy to differentiate proxy settings in different Elastic Agent groups. Global settings are not flexible enough in real-life scenarios.

The request is for Elastic Agent _including_ Elastic Endpoint.

**Describe a specific use case for the enhancement or feature:**

Elastic Agent deployment behind proxy in Fleet mode

**Comment by ferullo** [comment]
To whomever address this request: please inform Endpoint team so we can make changes too.

**Comment by nimarezainia** [comment]
@mostlyjason and @joshdover,   when we have the option for output definition per policy, the proxy configuration would also be applicable - correct?  currently applied at the global fleet output setting

**Comment by joshdover** [comment]
> when we have the option for output definition per policy, the proxy configuration would also be applicable - correct? currently applied at the global fleet output setting

@nchaulet can confirm, but I believe all output options will be available.

**Comment by nchaulet** [comment]
> @nchaulet can confirm, but I believe all output options will be available.

It's the same data model and same UI/API for global and per policy output so if one hase the option the other will have too.

**Comment by craiglawson** [comment]
> **Describe the enhancement:**
> 
> Agent connects to 3 elements: artifacts, Fleet Server and Elasticsearch. Today Fleet allows to specify proxy settings in global Fleet settings (Fleet UI) for Elasticsearch output only. To fully support proxy in Fleet mode, Fleet UI should allow to specify proxy settings for _all_ agent connections (not only the ES). Furthermore Fleet should allow to specify proxy settings per agent policy to differentiate proxy settings in different Elastic Agent groups. Global settings are not flexible enough in real-life scenarios.
> 
> The request is for Elastic Agent _including_ Elastic Endpoint.
> 
> **Describe a specific use case for the enhancement or feature:**
> 
> Elastic Agent deployment behind proxy in Fleet mode

This issue was raised on our behalf by @gbanasiak, we have the client where we provide a single deployment in ECE, they have multiple sites with a dedicated proxy at each so a global proxy config was not possible, we have EA deployed (with local changes to Linux services/registry), but we would like to provide endpoint protection to this client also, we could do this previously with Endgame (multiple sensor configs) but we are trying to migrate away from this for new deployments.

They have had to revert to their previous endpoint protection as we have been unable to modify proxy settings for ElasticEndpoint on a per-agent basis.

**Comment by jlind23** [comment]
@nimarezainia is this something we should start looking at rather soonish?

**Comment by mostlyjason** [comment]
@craiglawson We have several ways to set a proxy at the per-agent level here https://www.elastic.co/guide/en/fleet/7.17/fleet-agent-proxy-support.html. They work for all agent connections including artifacts, FS, and ES. Is there any reason this won't work for you? Also, is there a reason it needs to be set in the UI rather than on the agent side?

**Comment by craiglawson** [comment]
Yes, we can set per agent config on EA, I think our support case accelerated the current docs to be updated/published.

In my experience, this does not work for Endpoint Security, specifically Windows, we have a case with support (#00801509) which has been through this in detail.

**Comment by gbanasiak** [comment]
> For Fleet-managed agents, you can set the proxy using environment variables on the host.

That is not true for Endpoint Security - environment variable option is not available there. This means that if you have Endpoint Security deployed in endpoints running in different environments (some with proxy, some without), there is no way to handle it in a single Fleet deployment.

cc @nimarezainia
